azuread/table_azuread_service_principal.go

```diff
@@ -0,0 +1,122 @@
+			{Name: "owner_ids", Type: proto.ColumnType_JSON, Hydrate: getAdServicePrincipalOwners, Transform: transform.FromValue(), Description: "Id of the owners of the application. The owners are a set of non-admin users who are allowed to modify this object."},
+			{Name: "password_credentials", Type: proto.ColumnType_JSON, Description: "Represents a password credential associated with a service principal."},
+			{Name: "published_permission_scopes", Type: proto.ColumnType_JSON, Description: "The published permission scopes."},
+			{Name: "reply_urls", Type: proto.ColumnType_JSON, Description: "The URLs that user tokens are sent to for sign in with the associated application, or the redirect URIs that OAuth 2.0 authorization codes and access tokens are sent to for the associated application."},
+			{Name: "service_principal_names", Type: proto.ColumnType_JSON, Description: "Contains the list of identifiersUris, copied over from the associated application. Additional values can be added to hybrid applications. These values can be used to identify the permissions exposed by this app within Azure AD."},
+			{Name: "tags_src", Type: proto.ColumnType_JSON, Description: "Custom strings that can be used to categorize and identify the service principal.", Transform: transform.FromField("Tags"),},
+			{Name: "verified_publisher", Type: proto.ColumnType_JSON, Description: "Specifies the verified publisher of the application which this service principal represents."},
+
+			// // Standard columns
+			{Name: "tags", Type: proto.ColumnType_JSON, Description: ColumnDescriptionTags, Transform: transform.From(servicePrincipalTags),},
+			{Name: "title", Type: proto.ColumnType_STRING, Description: ColumnDescriptionTitle, Transform: transform.FromField("DisplayName", "ID")},
+			{Name: "tenant_id", Type: proto.ColumnType_STRING, Description: ColumnDescriptionTenant, Hydrate: plugin.HydrateFunc(getTenantId).WithCache(), Transform: transform.FromValue()},
+		},
+	}
+}
+
+//// LIST FUNCTION
+
+func listAdServicePrincipals(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
+	session, err := GetNewSession(ctx, d)
+	if err != nil {
+		return nil, err
+	}
+
+	client := msgraph.NewServicePrincipalsClient(session.TenantID)
+	client.BaseClient.Authorizer = session.Authorizer
+
+	// TODO filters
+	input := odata.Query{}
+	filter := ""
+	input.Filter = filter
+
+	servicePrincipals, _, err := client.List(ctx, input)
+	if err != nil {
+		if isNotFoundError(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	for _, servicePrincipal := range *servicePrincipals {
+		d.StreamListItem(ctx, servicePrincipal)
+	}
+
+	return nil, err
+}
+
+// Hydrate Functions
+
+// we didn't add the get function as it retries 5 times on 404 errors
```

Comment:
We have a get function in the `azure` plugin implementation (https://github.com/turbot/steampipe-plugin-azure/blob/main/azure/table_azure_ad_service_principal.go#L177), why are we unable to have it here?

Reply:
the library used here retries 5 times on a 404 error here, so we generally avoiding those APIs https://github.com/manicminer/hamilton/blob/main/msgraph/serviceprincipals.go#L91